### PR TITLE
EZP-23524: Implement field filters for Legacy Search Engine

### PR DESCRIPTION
**Pull request by pspanja**
This PR resolves https://jira.ez.no/browse/EZP-23524

#### What are field filters?

Basically, a second parameter of the [`SearchService::find()`](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/API/Repository/SearchService.php#L50) method. That is similar to what is available on some of the `ContentService:loadContent*` methods, which optionally accept a list of languages and always available indicator.

Contrary to the requirements for loading Content, in search implementation field filters conditions need to be added to the search backend query. That is needed in order to limit matching to specific languages and avoid `NotFound` exceptions when loading matched Content in `SearchService`.

#### Implementation

Ideally implementation would cover following criteria:

1. Field
2. MapLocationDistance
3. FieldRelation
4. FullText

However field filters support is here implemented only for `Field` and `MapLocationDistance` criteria.

`FieldRelation` can't support it as the required data does not exist in `ezcontentobject_link` table. In addition to `contentclassattribute_id` column that would require `contentobjectattribute_id`, so that `ezcontentobject_attribute` can be joined and field filter conditions applied.

`FullText` criterion can't support it as it relies on Legacy Stack's `eZSearchEngine` implementation, which does not support language filtering with fulltext search.

### CODE REVIEW

**Comment by pspanja** [comment]
Ready for review, ping @andrerom, @bdunogier, @glye, @lolautruche, @yannickroger

Note: this obviously will not be very performant. If we could update `ezcontentobject_attribute` with additional columns so that condition can be rewritten with `IN` operator, it should perform better as indexes could be employed in that case.

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/ContentTypeIdentifier.php [hunk 1]
```diff
@@ -13,13 +13,38 @@
...
 use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
 use eZ\Publish\Core\Persistence\Database\SelectQuery;
+use eZ\Publish\Core\Persistence\Database\DatabaseHandler;
+use eZ\Publish\SPI\Persistence\Content\Type\Handler as ContentTypeHandler;
 
 /**
  * Content type criterion handler
  */
-class ContentTypeIdentifier extends FieldBase
+class ContentTypeIdentifier extends CriterionHandler
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/ContentTypeIdentifier.php]: why this change?

Comment (pspanja) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/ContentTypeIdentifier.php]: Yes, it just made sense as we don't need Language Handler or field filters there. I initially added it for common dependency, but `ContentTypeIdentifier` is really not a type of `Field` criterion like others, so should not have been made to inherit from `FieldBase` in the first place.

Comment (andrerom) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/ContentTypeIdentifier.php]: k

**Comment by andrerom** [comment]
I'm mostly ok with this, so I'm going to add a +1
What I'm against is maybe rather fieldFilters itself and how it is (not) modeled, but maybe more a discussion to take later as a followup.

**Comment by bdunogier** [comment]
> FullText criterion can't support it as it relies on Legacy Stack's eZSearchEngine implementation, which does not support language filtering with fulltext search.

Does this mean that FullText search on Legacy Search will always match all languages ?

**Comment by pspanja** [comment]
@bdunogier 

> Does this mean that FullText search on Legacy Search will always match all languages ?

Yes. That is also how full text search behaves in LS with `ezsearchengine`.

File: eZ/Publish/API/Repository/Tests/SearchServiceFieldFiltersTest.php [hunk 2]
```diff
@@ -29,14 +29,34 @@ public function setUp()
...
+    {
+        if ( ltrim( get_class( $this->getSetupFactory() ), '\\' ) === 'eZ\\Publish\\API\\Repository\\Tests\\SetupFactory\\Legacy' )
+        {
+            $this->markTestSkipped(
+                "Legacy Search Engine does not support field filters with Fulltext criterion"
+            );
+        }
+    }
+
+    protected function supportsCustomFields()
```
Comment (bdunogier) [inline eZ/Publish/API/Repository/Tests/SearchServiceFieldFiltersTest.php]: I'd rename this to checkCustomFieldSupport(). "supports" kind of indicates that it will return a boolean, not skip the test if it doesn't.

File: eZ/Publish/API/Repository/Tests/SearchServiceFieldFiltersTest.php [hunk 3]
```diff
@@ -29,14 +29,34 @@ public function setUp()
...
+        if ( $setupFactory instanceof LegacySolr )
         {
-            $this->markTestIncomplete( "ATM implemented only for Elasticsearch storage" );
+            $this->markTestIncomplete( "Not implemented for Solr Search Engine" );
         }
 
         parent::setUp();
     }
 
+    protected function supportsFilteringFullText()
```
Comment (bdunogier) [inline eZ/Publish/API/Repository/Tests/SearchServiceFieldFiltersTest.php]: Same remark than below for the equivalent Custom field thing.

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FieldBase.php [hunk 4]
```diff
@@ -30,11 +32,124 @@
...
+        foreach ( $fieldFilters["languages"] as $languageCode )
+        {
+            $mask |= $this->languageHandler->loadByLanguageCode( $languageCode )->id;
+        }
+
+        return $mask;
+    }
+
+    /**
+     * Adds filed filters condition to the WHERE clause of the given $query.
```
Comment (lolautruche) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FieldBase.php]: Typo: s/filed/field

**Comment by lolautruche** [comment]
+1

File: eZ/Publish/API/Repository/Tests/SearchServiceFieldFiltersTest.php (hunk 2 above)
Comment (pspanja) [inline eZ/Publish/API/Repository/Tests/SearchServiceFieldFiltersTest.php]: Yup, thanks. Changed in https://github.com/ezsystems/ezpublish-kernel/commit/6778d580e834b26bdfabe37b6af85dea595fcb76.

File: eZ/Publish/API/Repository/Tests/SearchServiceFieldFiltersTest.php (hunk 3 above)
Comment (pspanja) [inline eZ/Publish/API/Repository/Tests/SearchServiceFieldFiltersTest.php]: Changed in https://github.com/ezsystems/ezpublish-kernel/commit/6778d580e834b26bdfabe37b6af85dea595fcb76.

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FieldBase.php (hunk 4 above)
Comment (pspanja) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/FieldBase.php]: Fixed in https://github.com/ezsystems/ezpublish-kernel/commit/c4d33cd4dd6f289dbc9a551c9ce96edf7e4aceb1.

**Comment by pspanja** [comment]
Thanks for the review, merging when Travis confirms.

**Comment by andrerom** [comment]
+1 *(travis confirms ;))*
